File: src/test/resources/PalindromeList.java
import java.util.*;
import org.algohub.engine.type.*;

public class Solution {
    /**
     * @param head Linked List to check if it's palindrome
     * @return  Indicates if input linked list is palindrome
     */
    public boolean isPalindrome(LinkedListNode<Integer> head) {
        LinkedListNode<Integer> reversed = reverseAndClone(head);
        return isEqual(head, reversed);
    }

    private LinkedListNode<Integer> reverseAndClone(LinkedListNode<Integer> node) {
        LinkedListNode<Integer> head = null;
        while (node != null) {
            LinkedListNode<Integer> n = new LinkedListNode<>(node.value);
            n.next = head;
            head = n;
            node = node.next;
        }
        return head;
    }

    private boolean isEqual(LinkedListNode<Integer> one, LinkedListNode<Integer> two) {
        while ((one != null && !one.isNull()) && (two != null && !two.isNull())) {
            if (one.value != two.value) {
                return false;
            }
            one = one.next;
            two = two.next;
        }
        return (one == null || one.isNull()) && (two == null || two.isNull());
    }
}

File: src/test/resources/PalindromeList.rb
class Solution
  def isPalindrome(head)
    reversed = reverse_and_clone(head)
    is_equal(head, reversed)
  end

  private

  def reverse_and_clone(node)
    head = nil

    while node
      n = com.jalgoarena.type.ListNode.new(node.value)
      n.next = head
      head = n
      node = node.next
    end

    head
  end

  def is_equal(one, two)
    while one && two
      return false if one.value != two.value

      one = one.next
      two = two.next
    end

    !one && !two
  end
end
